### ISSUE TITLE
Cross-platform use of platform specific packages

### ISSUE THREAD

**Issue by andschwa** [dotnet/corefx#6743]
Per the [.NET Standard Applications FAQ](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-standard-applications.md):

> Q: Can a .NET Standard application depend on platform specific packages like `Microsoft.Win32.Registry`?
> A: Yes, but it will only run on the platforms that support those packages.

This makes porting difficult, especially when your application supports features on a best-effort basis. I suggest the following answer:

> A: Yes, but it will throw `PlatformNotSupported` exceptions when used on platforms that do not support those packages.

Since this is not the case, .NET Core consumers are *unable* to write (or port) cross-platform applications that utilize platform-specific features.

This has been an ongoing discussion in dotnet/cli#1043.

@ericstj and I have a work-around, a `Microsoft.NETCore.Unsupported.Platform.API` package that provides runtime implementations that Do What I Mean (DWIM) on unsupported platforms (throw `PlatformNotSupported`). Using this, .NET applications that are ported from FullCLR to CoreCLR on multiple platforms do not have to eliminate their use of platform-specific features (such as the registry), but instead can continue to support their current behavior on Windows, and change behaviors on Linux (gracefully continue after a failed feature detection).

Consumers should be able to choose whether or not to use platform-specific features through their own feature-detection. This ability is integral to our internal project, and is currently a blocking obstacle.

**Comment by ericstj** [comment]
@andschwa can you share some data behind why cross-compiling won't work for you?  If you had two projects you could have the windows specific project add a runtimes section to specify it only builds for windows.

Alternatively there is a hack available today:
Package A > Package B > Platform-specific Package C
In A you can suppress compile for C ("exclude": "compile") permitting A to restore & publish for platforms where C is not supported.  This is how we are dealing with the problem with MSBuild & registry.

If we do provide mock assemblies for these packages we wouldn't want folks to take a dependency on them in their own nupkgs.  Doing so would expose other apps/libraries to runtime failures if they happened to use the platform specific API themselves.  Ideally it'd just be something an App could use to enable the platform specific packages.

**Comment by andschwa** [comment]
I have to speak in generalities since we're under NDA, but I'll do my best.

> can you share some data behind why cross-compiling won't work for you?

I'm not sure what you mean by cross-compiling not working for us. This term is overloaded. We'll never have a single cross-compiled binary that works on every platform. But we desire a project that is capable of being built for each of our supported platforms.

> If you had two projects you could have the windows specific project add a runtimes section to specify it only builds for windows.

We have the top level project, the host (A). This host is cross-platform, uses only C#, has a `Main` function, two frameworks (`netstandardapp1.5` for CoreCLR builds and `net451` for FullCLR builds), a set of runtimes which it supports (Ubuntu, CentOS, OS X, and Windows), and emits an executable.

We have a second project, the library (B) (in reality we have multiple libraries, but we'll stick with one here).  The host depends on the library (A -> B). This library is a massive legacy C# code base originally written for .NET 4.5. This library must continue to be compiled for .NET 4.5, and so uses a compile-time definition (in the frameworks section) `CORECLR` to differentiate between full and core builds.

This library has been ported to Linux (Ubuntu, CentOS, and OS X specifically, but will move to more). It uses a compile-time definition (in the configurations section, where we added a "Linux" configuration, since we cannot define this per runtime, yet) `LINUX` for the *few* places we absolutely must change code at compile time (e.g. the name of a library for a `DllImport` statement). However, because this is a single code base shared with the full and core Windows builds, use of Windows-specific features are guarded at runtime. There are many instances of nested calls represented by this pseudocode:

```c#
string get_config(RegistryKey k) {
    return get_real_config(RegistryKey k)
}

string get_real_config(RegistryKey k) {
    if (!Platform.IsWindows()) {
        return String.Empty;
    }
    # DANGER!
    return Registry.OpenKey(k);
}
```

The original caller of `get_config` is happy with an empty string; however, the `RegistryKey` type is still used in the function calls, even though the actual use of the registry on an unsupported platform is avoided.

So, this project is *not* Windows specific. However, when it is built for Windows, it uses Windows-specific features. Thus (B) takes a dependency on a platform-specific package (C), and B -> C is true on all platforms.

> In A you can suppress compile for C ("exclude": "compile") permitting A to restore & publish for platforms where C is not supported.

In this case, we do not compile C. We compile B, which uses types that are provided by C. On Windows, it also invokes methods in C; but on Linux, these methods are not invoked (and if they were, we would want to catch `PlatformNotSupportedException` and continue).

> This is how we are dealing with the problem with MSBuild & registry.

I'd like to know more about how you're dealing with this. I'll explore any alternative that is suggested. I don't quite understand what you mean by a) cross-compiling; b) two projects instead of one (I actually have 10); and c) suppressing compilation of (C).

> Ideally it'd just be something an App could use to enable the platform specific packages.

Absolutely agree, but it *seems* to be our only viable work-around due to .NET's architecture when dealing with platform-specific libraries.

**Comment by ericstj** [comment]
Fixed by #6988

**Comment by andschwa** [comment]
This is working great, thanks!
